test/Microsoft.AspNet.Mvc.IntegrationTests/CollectionModelBinderIntegrationTest.cs

```diff
@@ -536,5 +539,63 @@ private class Person2
             Assert.Equal(0, modelState.ErrorCount);
             Assert.True(modelState.IsValid);
         }
+
+        private class Person4
+        {
+            public IList<Address4> Addresses { get; set; }
+        }
+
+        private class Address4
+        {
+            public int Zip { get; set; }
+
+            public string Street { get; set; }
+        }
+
+        [Fact(Skip = "Extra ModelState key because of #2446")]
+        public async Task CollectionModelBinder_UsesCustomIndexes()
+        {
+            // Arrange
+            var argumentBinder = ModelBindingTestHelper.GetArgumentBinder();
+            var parameter = new ParameterDescriptor()
+            {
+                Name = "parameter",
+                ParameterType = typeof(Person4)
+            };
+
+            var operationContext = ModelBindingTestHelper.GetOperationBindingContext(request =>
+            {
+                var formCollection = new FormCollection(new Dictionary<string, string[]>()
+                {
+                    { "Addresses.index", new [] { "Key1", "Key2" } },
+                    { "Addresses[Key1].Street", new [] { "Street1" } },
+                    { "Addresses[Key2].Street", new [] { "Street2" } },
+                });
+
+                request.Form = formCollection;
+                request.ContentType = "application/x-www-form-urlencoded";
+            });
+
+            var modelState = new ModelStateDictionary();
+
+            // Act
+            var modelBindingResult = await argumentBinder.BindModelAsync(parameter, modelState, operationContext);
```

Comment:
I'm going to make a suggestion now that you'll hate me for. I deserve it probably.

I think that all of these integration tests should be updated to include verification for the validation nodes. 



Now that you hate me, I suggest you and I split up the work between us and do this after your wave of in-flight PRs is checked in.